### PR TITLE
Use Attribute Converter for mapParams

### LINKED ISSUE #446: Change map params to be represented as a map and not string

### PR DESCRIPTION
**Pull request by Sheikah45**
Fixes #446

### CODE REVIEW

**Review by bukajsytlos — APPROVED** [review]

**Review by bukajsytlos — CHANGES_REQUESTED** [review]
Propably we should also do tests for read.
Also write tests should have asserts whether it was correctly writen

**Comment by Sheikah45** [comment]
tests added

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import lombok.extern.slf4j.Slf4j;
+
+import javax.persistence.AttributeConverter;
+import javax.persistence.Converter;
+import java.io.IOException;
+import java.util.Map;
+
+@Slf4j
+@Converter
+public class MapParamsConverter implements AttributeConverter<Map<String, Object>, String> {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: This is not a MapParamsConverter. This is a reusable JsonConverter.

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  private final ObjectMapper objectMapper;
+
+  public MapParamsConverter(ObjectMapper objectMapper) {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: use @RequiredArgsConstructor instead

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  @Override
+  public String convertToDatabaseColumn(Map<String, Object> mapParams) {
+
+    String mapParamsJson = null;
+    try {
+      mapParamsJson = objectMapper.writeValueAsString(mapParams);
+    } catch (final JsonProcessingException e) {
+      log.error("JSON writing error", e);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: "Failed to convert Json object to String"

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+public class MapParamsConverter implements AttributeConverter<Map<String, Object>, String> {
+
+  private final ObjectMapper objectMapper;
+
+  public MapParamsConverter(ObjectMapper objectMapper) {
+    this.objectMapper = objectMapper;
+  }
+
+  @Override
+  public String convertToDatabaseColumn(Map<String, Object> mapParams) {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: call the parameter jsonPayload instead

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      mapParamsJson = objectMapper.writeValueAsString(mapParams);
+    } catch (final JsonProcessingException e) {
+      log.error("JSON writing error", e);
+    }
+
+    return mapParamsJson;
+  }
+
+  @Override
+  public Map<String, Object> convertToEntityAttribute(String mapParamsJSON) {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: rename mapParamsJSON to jsonAsString

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    Map<String, Object> mapParams = null;
+    if (mapParamsJSON != null) {
+      try {
+        mapParams = objectMapper.readValue(mapParamsJSON, Map.class);
+      } catch (final IOException e) {
+        log.error("JSON reading error", e);
+      }
+    }
+
+    return mapParams;
```
Comment (Brutus5000) [inline src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: rename mapParams to jsonPayload

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  public String convertToDatabaseColumn(Map<String, Object> mapParams) {
+
+    String mapParamsJson = null;
+    try {
+      mapParamsJson = objectMapper.writeValueAsString(mapParams);
+    } catch (final JsonProcessingException e) {
+      log.error("JSON writing error", e);
+    }
+
+    return mapParamsJson;
```
Comment (Brutus5000) [inline src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: rename mapParamsJson to jsonAsString

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Override
+  public Map<String, Object> convertToEntityAttribute(String mapParamsJSON) {
+
+    Map<String, Object> mapParams = null;
+    if (mapParamsJSON != null) {
+      try {
+        mapParams = objectMapper.readValue(mapParamsJSON, Map.class);
+      } catch (final IOException e) {
+        log.error("JSON reading error", e);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: "Faild to read stringified Json"

**Review by Brutus5000 — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java (hunk 8 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: Done

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java (hunk 7 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: Done

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java (hunk 6 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: Done

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java (hunk 5 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: Done

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java (hunk 4 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: Done

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java (hunk 3 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: Done

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java (hunk 2 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: Done

File: src/main/java/com/faforever/api/data/converter/MapParamsConverter.java (hunk 1 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/api/data/converter/MapParamsConverter.java]: Done

**Pull request merged**